olive/common/ort_inference.py

```diff
@@ -51,26 +75,299 @@ def get_ort_inference_session(
         for key, value in extra_session_config.items():
             sess_options.add_session_config_entry(key, value)
 
-    if isinstance(execution_provider, str):
-        execution_provider = [execution_provider]
-    else:
-        # execution providers should be list[str]
-        assert isinstance(execution_provider, list)
-        assert all(isinstance(ep, str) for ep in execution_provider)
-
-    for idx, ep in enumerate(execution_provider):
-        if ep == "QNNExecutionProvider":
+    # execution providers and provider options
+    providers, provider_options = check_and_normalize_provider_args(
+        inference_settings.get("execution_provider"),
+        inference_settings.get("provider_options"),
+        ort.get_available_providers(),
+    )
+    for idx, provider in enumerate(providers):
+        if provider in ["CUDAExecutionProvider", "DmlExecutionProvider"] and device_id is not None:
```

Comment:
Will Rocm/Migraphx also need device_id? I also considered to add the device id, but just found another simple solution by using `CUDA_VISIBLE_DEVICES` out of Olive.
![image](https://github.com/microsoft/Olive/assets/13343117/5706c327-56cc-403b-a4e3-545f69cdb4ff)

If you have strong requirements to use `device_id`, could you help to update the related docs later?

Reply:
Yeah, we use `CUDA_VISIBLE_DEVICES` for regular onnx model. This part is added for distributed onnx models where there are multiple models that need to be on their own device. 
This code change doesn't introduce a new behavior. I am only aggregating the logic in `ONNXModelHandler.prepare_session` into this module https://github.com/microsoft/Olive/blob/683ed65a5c11c614759d186c607bcc26553afae3/olive/model/handler/onnx.py#L104 

So, this is only used internally by Olive and is not really user facing. Most of the time, `device_id` here will be `None`. Same for `OrtInferenceSession` wrapper.

Reply:
> So, this is only used internally by Olive and is not really user facing. Most of the time, device_id here will be None. Same for OrtInferenceSession wrapper.
Got it. Could you help also update the code comments to tell?

Reply:
sure, I will add some comments above this to describe the current use cases.

Reply:
Done!